MFT Noise calibration

mcoquet642 opened the pull request:
Implementing noise scan calibration for MFT following the lines of ITS writing a single NoiseMap object in the CCDB. Additional features include an option **--useDigits** to make the device run in mode based on digits rather than clusters (default mode).

MauriceCoke committed:
First version of MFT calibration : generating noise map and writing in CCDB

mcoquet642 committed:
Update README.md

MauriceCoke committed:
modifié :         Detectors/ITSMFT/MFT/calibration/CMakeLists.txt
	nouveau fichier : Detectors/ITSMFT/MFT/calibration/include/MFTCalibration/NoiseCalibratorDigits.h
	nouveau fichier : Detectors/ITSMFT/MFT/calibration/include/MFTCalibration/NoiseCalibratorDigitsSpec.h
	modifié :         Detectors/ITSMFT/MFT/calibration/src/MFTCalibrationLinkDef.h
	nouveau fichier : Detectors/ITSMFT/MFT/calibration/src/NoiseCalibratorDigits.cxx
	nouveau fichier : Detectors/ITSMFT/MFT/calibration/src/NoiseCalibratorDigitsSpec.cxx
	modifié :         Detectors/ITSMFT/MFT/calibration/testWorkflow/README.md
	modifié :         Detectors/ITSMFT/MFT/calibration/testWorkflow/mft-calib-workflow.cxx

MauriceCoke committed:
modifié :         Detectors/ITSMFT/MFT/calibration/testWorkflow/mft-calib-workflow.cxx

MauriceCoke committed:
Implementation of noise calibration DPL using digits

mcoquet642 committed:
Update README.md

MauriceCoke committed:
TimeSlot calibration

MauriceCoke committed:
Clang format

mcoquet642 committed:
Fix typo

mcoquet642 committed:
Fix typo

mcoquet642 committed:
Fix typo

mcoquet642 committed:
Update NoiseCalibrator.h

mcoquet642 committed:
Update NoiseCalibratorDigits.h

MauriceCoke committed:
Updating TimeSolt implementation

MauriceCoke committed:
Clang format

MauriceCoke committed:
Digits/Clusters implementation

MauriceCoke committed:
Using StringUtils

MauriceCoke committed:
Clang Format

MauriceCoke committed:
Fixing spaces and tabs

shahor02 commented on Detectors/ITSMFT/MFT/calibration/src/NoiseCalibratorSpec.cxx:
```diff
@@ -0,0 +1,181 @@
+            << " : " << info.getEndValidityTimestamp();
+
+  using clbUtils = o2::calibration::Utils;
+  output.snapshot(
+    Output{clbUtils::gDataOriginCLB, clbUtils::gDataDescriptionCLBPayload, 0}, *image.get());
+  output.snapshot(
+    Output{clbUtils::gDataOriginCLB, clbUtils::gDataDescriptionCLBInfo, 0}, info);
```
Please note that the convention for the messages sent to CCDB populator was changed since https://github.com/AliceO2Group/AliceO2/pull/6150 in order to avoid clashes between the inputs from different calibrators. The documentation was not fully updated, just pushed https://github.com/AliceO2Group/AliceO2/pull/6234.
Now it should be  e.g.
````
output.snapshot(Output{clbUtils::gDataOriginCDBPayload, "MFT_NoiseMap", 0}, *image.get());
output.snapshot(Output{clbUtils::gDataOriginCDBWrapper, "MFT_NoiseMap", 0}, info);
````

mcoquet642 replied:
Hi Ruben, thanks for the comments. This is now corrected.

shahor02 commented on Detectors/ITSMFT/MFT/calibration/src/NoiseCalibratorSpec.cxx:
```diff
@@ -0,0 +1,181 @@
+  using clbUtils = o2::calibration::Utils;
+  std::vector<OutputSpec> outputs;
+  outputs.emplace_back(
+    ConcreteDataTypeMatcher{clbUtils::gDataOriginCLB, clbUtils::gDataDescriptionCLBPayload});
+  outputs.emplace_back(
+    ConcreteDataTypeMatcher{clbUtils::gDataOriginCLB, clbUtils::gDataDescriptionCLBInfo});
```
Respectively, in the getNoiseCalibratorSpec the outputs should be declared as 
````
outputs.emplace_back(ConcreteDataTypeMatcher{clbUtils::gDataOriginCDBPayload, "MFT_NoiseMap"});
outputs.emplace_back(ConcreteDataTypeMatcher{clbUtils::gDataOriginCDBWrapper, "MFT_NoiseMap"});
````

shahor02 commented on Detectors/ITSMFT/MFT/calibration/src/NoiseCalibratorSpec.cxx:
```diff
@@ -0,0 +1,181 @@
+    outputs,
+    AlgorithmSpec{adaptFromTask<NoiseCalibratorSpec>(useDigits)},
+    Options{
+      {"1pix-only", VariantType::Bool, false, {"Fast 1-pixel calibration only"}},
```
What is the purpose of this option? It will miss the noisy channels which happen to be neighbors (e.g. noisy column).

mcoquet642 replied:
This was a remanent of what was in ITS code, now removed.

shahor02 commented on Detectors/ITSMFT/MFT/calibration/src/NoiseSlotCalibrator.cxx:
```diff
@@ -0,0 +1,147 @@
+bool NoiseSlotCalibrator::processTimeFrame(gsl::span<const o2::itsmft::Digit> const& digits,
+                                           gsl::span<const o2::itsmft::ROFRecord> const& rofs)
+{
+  calibration::TFType nTF = rofs[0].getBCData().orbit / mHBFperTF;
```
The calibration may start from any orbit, so the nTF in general will not reflect the TFID. In other time-slot-based calibrations we are using `auto tfcounter = o2::header::get<o2::framework::DataProcessingHeader*>(pc.inputs().get("input").header)->startTime;`. This is also not really correct, but at least, once we come to agreement on the common TF counter definition, we can change all centrally. I would suggest to use the same: extract the TF counter in the CalibrationSpec and pass it to the Calibrator, e.g. like in `https://github.com/AliceO2Group/AliceO2/blob/c2f765d42f324beb1e1ecda9cf5ba4acf29704cd/Detectors/TOF/calibration/testWorkflow/LHCClockCalibratorSpec.h#L51-L54`
Then, you also will not need the `mHBFperTF`.
Actually, for you the specific slot should be irrelevant, since you have just 1 with 0 - infinity span.

mcoquet642 replied:
Thanks for pointing this out. Indeed I had some troubles with this. I think the same line is in the ITS code.

shahor02 commented on Detectors/ITSMFT/MFT/calibration/src/NoiseCalibratorSpec.cxx:
```diff
@@ -0,0 +1,181 @@
+  mStart = ic.options().get<int64_t>("tstart");
+  mEnd = ic.options().get<int64_t>("tend");
+
+  mCalibrator = std::make_unique<CALIBRATOR>(onepix, probT, HBperTF);
```
You should set the length of the calibration slot, i.e. in your case infinite (e.g. `std::numeric_limits<TFType>::max`) since you have just 1 slot.

mcoquet642 replied:
Just a comment : by default this PR doesn't use the NoiseSlotCalibrator class but the NoiseCalibrator which doesn't use TimeSlots, as Jouri mentioned during the discussion that in the end ITS doesn't use the former. So I included the appropriate line but commented it out.

shahor02 reviewed (changes requested):
Hi @mcoquet642 , please see some comments below.
This version of time-slot based calibration is essentially similar to that of ITS, where the slot receives the digits or clusters fills the map. As such, it can aggregate the data both from the EPNs and from FLPs (if the decoding/clusterization is done there).
An alternative, as we discussed at the meeting, might be the FLP processing which builds the map for the chips seen on each FLP, then the time-slot-calibration receives the maps from different FLPs, merges them and sends to the CCDB populator.

MauriceCoke committed:
Changes for PR

	modifié :         ../include/MFTCalibration/NoiseCalibrator.h
	modifié :         ../include/MFTCalibration/NoiseCalibratorSpec.h
	modifié :         ../include/MFTCalibration/NoiseSlotCalibrator.h
	modifié :         NoiseCalibrator.cxx
	modifié :         NoiseCalibratorSpec.cxx
	modifié :         NoiseSlotCalibrator.cxx

MauriceCoke committed:
Clang format

shahor02 commented on Detectors/ITSMFT/MFT/calibration/src/NoiseCalibratorSpec.cxx:
```diff
@@ -0,0 +1,174 @@
+  mEnd = ic.options().get<int64_t>("tend");
+
+  mCalibrator = std::make_unique<CALIBRATOR>(probT);
+  //  mCalibrator->setSlotLength(std::numeric_limits<TFType>::max); //For TimeSlot calibration
```
Since you never need multiple slots, why don't you set it in the NoiseSlotCalibrator constructor?

mcoquet642 replied:
Should be done now.

shahor02 commented on Detectors/ITSMFT/MFT/calibration/src/NoiseCalibratorSpec.cxx:
```diff
@@ -0,0 +1,174 @@
+  // fill meta map
+  for (auto& p : keyvalues) {
+    meta[p.first] = p.second;
+  }
```
Why don't you do directly:
````
std::map<std::string, std::string> meta;
auto toKeyValPairs = [&meta](std::vector<std::string> const& tokens) {
  for (auto&amp; token : tokens) {
     auto keyval = Str::tokenize(token, '=', false);
     if (keyval.size() != 2) {
        LOG(ERROR) << "Illegal command-line key/value string: " << token;
        continue;
     }
     Str::trim(keyval[1]);
     meta[keyval[0]] = keyval[1];
  }
};
toKeyValPairs(Str::tokenize(mMeta, ';', true));
````

mcoquet642 replied:
Thanks, indeed. Now fixed.

mcoquet642 committed:
Changes for PR

mcoquet642 committed:
Clang Format

mcoquet642 committed:
Fix typo

shahor02 reviewed (approved):
approving to trigger CI check

shahor02 commented:
Squashing/merging, CI failures are not related to the PR.

@TimoWilken all 3 required tests are regularly failing: macos does not start, in o2 the compiler is killed, and fullCI fails at random stages, I suspect due to the insufficient memory.
